### PR TITLE
[Build] Integrate build-tool plugin invocations into build planning

### PR DESCRIPTION
**Pull request by xedin**
Plugins were invoked in an ad-hoc way in `BuildOperation.plan`, a better place for handling their is during build plan construction.

### Motivation:

Build-tool plugins are logically coupled with modules that depend on them, so it makes sense to invoke them incrementally while build plan is forming build descriptions for such modules.

### Modifications:

This is mostly an NFC change which splits up `invokeBuildToolPlugins` and moves some of its functionality into a new `PluginModule.invoke` method and the remainder into the `BuildPlan.invokeBuildToolPlugins`. `runPrebuildCommands` is now part of a `BuildPlan` as well.

### Result:

Build-tool plugins are now integrated into build planning which as a side-effect would help tools like LSP to handle them better.

### CODE REVIEW

**Comment by xedin** [comment]
@MaxDesiatov I wanted to move `computePluginGeneratedFiles` from `ModulesGraph` but it's marked `public` at the moment, so I'm not sure if I can touch that or not...

File: Sources/Build/BuildOperation.swift [hunk 1]
```diff
@@ -734,69 +734,38 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
     private func plan(subset: BuildSubset? = nil) throws -> (description: BuildDescription, manifest: LLBuildManifest) {
         // Load the package graph.
         let graph = try getPackageGraph()
-        let buildToolPluginInvocationResults: [ResolvedModule.ID: (target: ResolvedModule, results: [BuildToolPluginInvocationResult])]
-        let prebuildCommandResults: [ResolvedModule.ID: [PrebuildCommandResult]]
-        // Invoke any build tool plugins in the graph to generate prebuild commands and build commands.
-        if let pluginConfiguration, !self.config.shouldSkipBuilding(for: .target) {
+
+        let pluginTools: [ResolvedModule.ID: [String: PluginTool]]
+        // FIXME: This is unfortunately but we need to build plugin tools upfront at the moment.
```
Comment (rauhul) [inline Sources/Build/BuildOperation.swift:739]: I think it would be helpful to future readers to explain why

Comment (xedin) [reply Sources/Build/BuildOperation.swift:739]: Sure, I will add a comment that explains it.

File: Sources/Build/BuildOperation.swift [hunk 2]
```diff
@@ -734,69 +734,38 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
...
-                    try self.runPrebuildCommands(for: $0.results)
-                }
-            })
         } else {
-            buildToolPluginInvocationResults = [:]
-            prebuildCommandResults = [:]
+            pluginTools = [:]
         }
 
+        // Create the build plan based, on the graph and any information from plugins.
```
Comment (MaxDesiatov) [inline Sources/Build/BuildOperation.swift:754]:
```suggestion
        // Create the build plan based on the modules graph and any information from plugins.
```

File: Sources/Build/BuildOperation.swift [hunk 3]
```diff
@@ -862,6 +797,48 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
         return (buildDescription, buildManifest)
     }
 
+    private func detectUnhandledFiles(
```
Comment (MaxDesiatov) [inline Sources/Build/BuildOperation.swift:800]: Missing a doc comment, I would like to see it elaborate on what "handling" means in this context.

```suggestion
    /// 
    private func detectUnhandledFiles(
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 4]
```diff
@@ -266,22 +267,24 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
...
         self.toolsBuildParameters = toolsBuildParameters
         self.graph = graph
-        self.buildToolPluginInvocationResults = buildToolPluginInvocationResults
-        self.prebuildCommandResults = prebuildCommandResults
         self.shouldDisableSandbox = disableSandbox
         self.fileSystem = fileSystem
         self.observabilityScope = observabilityScope.makeChildScope(description: "Build Plan")
 
+        var buildToolPluginInvocationResults: [ResolvedModule.ID: [BuildToolPluginInvocationResult]] = [:]
+        var prebuildCommandInvocationResults: [ResolvedModule.ID: [PrebuildCommandResult]] = [:]
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:286]: Let's stick to the naming used in the SE proposal: "command plugins".

```suggestion
        var commandPluginInvocationResults: [ResolvedModule.ID: [PrebuildCommandResult]] = [:]
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 5]
```diff
@@ -356,6 +359,35 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
...
+                    fileSystem: fileSystem,
+                    observabilityScope: observabilityScope,
+                    surfaceDiagnostics: true
+                )
+
+                if pluginInvocationResults.contains(where: { !$0.succeeded }) {
+                    throw StringError("build planning stopped due to build-tool plugin failures")
+                }
+
+                let prebuildCommandResults = try Self.runPrebuildCommands(
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:380]: ditto

```suggestion
                let commandPluginsResults = try Self.runPrebuildCommands(
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 6]
```diff
@@ -692,6 +727,203 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
...
+    /// that will later be run before or during the build, and can also emit debug output and diagnostics.
+    ///
+    /// Each result returned by this function includes an ordered list of commands to run before the build
+    /// of the module, and another list of the commands to incorporate into the build graph so they run
+    /// at the appropriate times during the build.
+    ///
+    /// Any warnings and errors related to running the plugin will be emitted to `diagnostics` when
+    /// `surfaceDiagnostics` parameter is set to `true`.
+    ///
+    /// Note that warnings emitted by the the plugin itself will be returned in the PluginEvaluationResult
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:743]: Looks like the comment is outdated, there's no such type as `PluginEvaluationResult` anywhere in the codebase. It's probably referring to `BuildToolPluginInvocationResult`?

```suggestion
    /// Note that warnings emitted by the the plugin itself will be returned in the ``BuildToolPluginInvocationResult``
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 7]
```diff
@@ -692,6 +727,203 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
...
+            // Add a BuildToolPluginInvocationResult to the mapping.
+            buildToolPluginResults.append(result)
+        }
+
+        return buildToolPluginResults
+    }
+
+    /// Runs any prebuild commands associated with the given list of plugin invocation results,
+    /// in order, and returns the results of running those prebuild commands.
+    fileprivate static func runPrebuildCommands(
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:871]:
```suggestion
    /// Runs any command plugins associated with the given list of plugin invocation results,
    /// in order, and returns the results of running those prebuild commands.
    fileprivate static func runCommandPlugins(
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 8]
```diff
@@ -692,6 +727,203 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
...
+        using pluginConfiguration: PluginConfiguration,
+        for pluginResults: [BuildToolPluginInvocationResult],
+        fileSystem: any FileSystem,
+        observabilityScope: ObservabilityScope
+    ) throws -> [PrebuildCommandResult] {
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:876]: Probably worth renaming the type too, to make it more obvious which feature it's referring to.

```suggestion
    ) throws -> [CommandPluginResult] {
```

File: Sources/Build/BuildPlan/BuildPlan.swift [hunk 9]
```diff
@@ -692,6 +727,203 @@ public class BuildPlan: SPMBuildCore.BuildPlan {
...
+        // Run through all the commands from all the plugin usages in the target.
+        try pluginResults.map { pluginResult in
+            // As we go we will collect a list of prebuild output directories whose contents should be input to the
+            // build,
+            // and a list of the files in those directories after running the commands.
```
Comment (MaxDesiatov) [inline Sources/Build/BuildPlan/BuildPlan.swift:881]: comment formatting nit
```suggestion
            // As we go we will collect a list of prebuild output directories whose contents should be input to the
            // build, and a list of the files in those directories after running the commands.
```

Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift:881]: Interestingly this is what `swiftformat` did...

Comment (MaxDesiatov) [reply Sources/Build/BuildPlan/BuildPlan.swift:881]: Yes, its comments line length reflow needs a second pass by a human 😅

File: Sources/Build/BuildOperation.swift (hunk 1 above)
Comment (xedin) [reply Sources/Build/BuildOperation.swift]: @rauhul how does this look?

File: Sources/Build/BuildOperation.swift (hunk 2 above)
Comment (xedin) [reply Sources/Build/BuildOperation.swift]: Thanks, done!

File: Sources/Build/BuildOperation.swift (hunk 3 above)
Comment (xedin) [reply Sources/Build/BuildOperation.swift]: Renamed to `diagnoseUnhandledFiles` and added a comment.

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 4 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: Renamed!

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 5 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: ✅

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 6 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: ✅

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 8 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: Renamed!

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 7 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: This was code that I moved from `BuildOperation`, fixed!

File: Sources/Build/BuildPlan/BuildPlan.swift (hunk 9 above)
Comment (xedin) [reply Sources/Build/BuildPlan/BuildPlan.swift]: ✅

**Review by MaxDesiatov — APPROVED** [review]
Thank you!

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by xedin** [comment]
@MaxDesiatov Test is failing because of

```
'objc[71462]: Class _TtC11SwiftSyntax16BumpPtrAllocator is implemented in both /Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/.build/x86_64-apple-macosx/debug/swift-package-manager (0x108d63050) and /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/libSwiftSyntax.dylib (0x10e6a67f8). One of the two will be used. Which one is undefined.
objc[71462]: Class _TtC11SwiftSyntax23SourceLocationConverter is implemented in both /Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/.build/x86_64-apple-macosx/debug/swift-package-manager (0x108d633f0) and /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/libSwiftSyntax.dylib (0x10e6a6be0). One of the two will be used. Which one is undefined.
```

Do you know if this was some kind of recent change?

**Comment by xedin** [comment]
Actually nevermind, I changed the output here, wonder why `./Utilities/bootstrap test` didn't fail on this...

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Windows platform

File: Sources/Build/BuildOperation.swift [hunk 10]
```diff
@@ -734,69 +734,41 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
     private func plan(subset: BuildSubset? = nil) throws -> (description: BuildDescription, manifest: LLBuildManifest) {
         // Load the package graph.
         let graph = try getPackageGraph()
-        let buildToolPluginInvocationResults: [ResolvedModule.ID: (target: ResolvedModule, results: [BuildToolPluginInvocationResult])]
-        let prebuildCommandResults: [ResolvedModule.ID: [PrebuildCommandResult]]
-        // Invoke any build tool plugins in the graph to generate prebuild commands and build commands.
-        if let pluginConfiguration, !self.config.shouldSkipBuilding(for: .target) {
+
+        let pluginTools: [ResolvedModule.ID: [String: PluginTool]]
+        // FIXME: This is unfortunate but we need to build plugin tools upfront at the moment because
```
Comment (rauhul) [inline Sources/Build/BuildOperation.swift:739]: Is this really the case? Don't build tool plugin have to declare the inputs and outputs statically compared to pre-build tools which need to be run in order to capture the output files?

Comment (xedin) [reply Sources/Build/BuildOperation.swift:739]: I am not sure I follow, we need to build tools in order to invoke plugins to form a complete manifest.

Comment (MaxDesiatov) [reply Sources/Build/BuildOperation.swift:739]: > Is this really the case? Don't build tool plugin have to declare the inputs and outputs statically compared to pre-build tools which need to be run in order to capture the output files?

IIUC the answer is "no", `createBuildCommands` is invoked during the execution of the plugin, I don't think that could count as "static".

**Pull request merged**